### ISSUE TITLE
Bazelify the GMP Precision Arithmetic Library

### ISSUE THREAD

**Issue by rauljordan** [prysmaticlabs/go-bls#2]
Hi all,

A critical component of this library is the [GNU Multiple Precision Arithmetic Library](https://gmplib.org/) written in C used for most underlying cryptographic calculations that are part of BLS. A major problem with this is that most architectures do not ship with GMP installed globally, and therefore any user wishing to use go-bls in their project will need to pollute their global namespace with libgmp.

We have opted for using the [Bazel](https://build.bazel) build system created by Google for dependency management and packaging of this repository for use in other projects. In particular, go-bls will be a critical component of the [Prysm](https://github.com/prysmaticlabs/prysm) client created by Prysmatic Labs for Ethereum 2.0. Prysm uses the Bazel build tool extensively and ideally we'd want to import go-bls via Bazel as well. The problem with this is that GMP is not included as a Bazel dependency of go-bls, and therefore the build will fail due not finding the library.

A solution to this would be to convert lib GMP into a Bazel project following the guidelines for C/C++ projects in the Bazel documentation [here](https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library). This is fairly extensive effort as GMP is a very big library containing multiple pieces. Tackling this issue would allow Prysm and any other Bazel-enabled project to import go-bls without requiring users to have GMP installed on their machine.

We are requesting to bounty this project out for motivated individuals to make this happen. This is a high priority for us at Prysmatic Labs.

### Starting Point

This issue created by @prestonvanloon is a great starting point for someone tackling this: [unable to gen_rule as part of http_archive](https://github.com/bazelbuild/bazel/issues/6344).

Preston began by using http_archive to pull the libgmp official repo and begin creating a bazel BUILD file for the system directly.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1000.0 DAI (1000.0 USD @ $1.0/DAI)  attached to it as part of the Ethereum Foundation fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $54,603.34 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 3 weeks, 1 day ago.
Please review their action plans below:


**1) [pepeinvest](https://gitcoin.co/profile/pepeinvest) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?mutate_worker_action=approve&worker=pepeinvest) | [reject worker](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?mutate_worker_action=reject&worker=pepeinvest))_.**

Creative Bussines developers And finance LTd

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).


**2) [pacamara](https://gitcoin.co/profile/pacamara) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?mutate_worker_action=approve&worker=pacamara) | [reject worker](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?mutate_worker_action=reject&worker=pacamara))_.**

Hi! I'd like to start work on this. Action plan: 1. Install Bazel. 2. Write appropriate rules for lib GMP. 3. Test on Ubuntu, OSX, and possibly Windows (if required).

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).


**3) [robin-thomas](https://gitcoin.co/profile/robin-thomas) has been approved to start work.**

* Download GMP tar file using `new_http_archive()` instead of `http_archive`
* Create the BUILD file to convert the library into a bazel repo
* Add the above dependency to `go-bls` library

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).

**Comment by spm32** [comment]
Hey @jakerockland you're approved but please let us know ASAP if you don't think you'll be able to do it :)

**Comment by samparsky** [comment]
@jakerockland If you want we can possibly work on together on this & maybe share the bounty?

**Comment by jakerockland** [comment]
@rauljordan @ceresstation To make sure I'm understanding the issue correctly, is the end deliverable for this something along the lines of a new repository containing the source for GMP, with a Bazel build file that allows the GMP libraries to be built with Bazel instead of Make?

**Comment by jakerockland** [comment]
@samparsky I'm open to collaborating if there is a sensible way to split things up, what did you have in mind?

**Comment by prestonvanloon** [comment]
@jakerockland Yes. This should be built entirely using Bazel as the build tool.

It should build on:
- x86_64
- Mac OS X
- Windows
- arm64
- And any other relevant cpu architecture.

**Comment by jakerockland** [comment]
@prestonvanloon Thank you for the clarification. Given the high priority of this and what else I have in the pipeline, after taking some time to scope this out last night and realize how much of a behemoth/custom configuration scripts there are with GMP and confirming now that I understood properly, I think that it may be better to let someone else take on this bounty @ceresstation.

I've spend a good amount of time today trying to get things to work with this approach, to no avail https://github.com/bazelbuild/rules_go/issues/1233. Have you seen this issue @prestonvanloon?

**Comment by spm32** [comment]
Fair @jakerockland, @samparsky are you still up for taking it on? Do you feel confident you're able to do it?

**Comment by samparsky** [comment]
@ceresstation Started work on it, will let you know my progress by tomorrow

**Comment by samparsky** [comment]
@jakerockland @ceresstation I have been able to get it to build successfully using Bazel on MacOSX. Currently testing for ubuntu using a docker image.
For windows its recommended in the gmplib docs (https://cs.nyu.edu/~exact/core/gmp/index.html) to use a Linux environment Cygwin or MinGW

**Comment by rauljordan** [comment]
That's awesome, thanks for the update @samparsky!

**Comment by samparsky** [comment]
@prestonvanloon I have been able to get it to build on using Bazel on Ubuntu & MacOSX. 

But I am facing an issue I am able to get the build to work properly on MacOSX with Bazel sandbox mode off only & on Linux it works with sandbox mode on only. 

I think it has something to do with the way the underlying filesystem works in these different OSes.

You can check out the details here 
https://github.com/samparsky/bazel-gmp

The only difference between the two folders is `local=0 or 1` in the `gmplib.BUILD` file. 
In the mac folder its `local=1` which means sandbox mode is off & in linux folder its `local=0` which means sandbox mode is on.

**Comment by prestonvanloon** [comment]
@samparsky I took a look at your repository. It looks like your implementation is a genrule calling configure and make. 

The goal of this task is to use bazel to build the library, not make. Additionally, you shouldn’t require users to install any system dependencies. They should be able to have a working bazel installation and run `bazel build //...` without additional steps. 

I realize this issue might have been misleading because I attempted to use the genrule approach as a bandaid fix for building go-bls without gmp installed on the system. Sorry about the confusion.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@samparsky due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?snooze=1) | [3 days](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?snooze=3) | [5 days](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?snooze=5) | [10 days](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?snooze=10) | [100 days](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782?snooze=100)

**Comment by samparsky** [comment]
@prestonvanloon I tried building it manually couldn't get it to work. Also emailed the GMPlib mailing list to get any pointers couldn't get any positive response

**Comment by robin-thomas** [comment]
@prestonvanloon If it's not an issue, I would like to work on it. As I understand, you need to download the GMP library, and use bazel to build it.

**Comment by prestonvanloon** [comment]
@robin-thomas sounds good to me. As long as bazel is the tool to actually build it. (Not using bazel as a wrapper for make or similar)

**Comment by prestonvanloon** [comment]
@ceresstation can you help transfer this issue to @robin-thomas please?

**Comment by robin-thomas** [comment]
@prestonvanloon I had already applied for the issue in gitcoin.
https://github.com/prysmaticlabs/go-bls/issues/2#issuecomment-439647266

Are you unable to approve me with that link?

**Comment by robin-thomas** [comment]
@prestonvanloon @rauljordan Could you take a look at: https://github.com/robin-thomas/bazelify-gmp?

Its still WIP, but to give you some idea.

**Comment by prestonvanloon** [comment]
Looking great @robin-thomas!

**Comment by robin-thomas** [comment]
@prestonvanloon Thanks! I'll continue working on it. Still loads more to be done.

@ceresstation Can you kindly transfer this issue to me? Thanks.

**Comment by robin-thomas** [comment]
@prestonvanloon, @rauljordan. Wanted to give you an update.

Many of the *.h, *.c files are generated on the fly by executing the configure script, many of which are not even known during the analysis stage of Bazel. Bazel requires these names before the build even starts! Related to how Bazel constructs its dependency graph.

Anyway, the only solution here is to zip all those files into one file with a fixed name that can be passed to Bazel. But then again, there is no native support to pass that zip file as the input to cc_* rules for compilation. The only way around it is using Tree Artifacts, and it's a bit of hack to get it working and compiled into a library.

The configure script also generates loads of *.asm files based on the host architecture. It *cannot* be passed to Bazel's cc_rules, as Bazel already assumes *.asm files to be assembly code generated by the compiler. https://github.com/bazelbuild/bazel/issues/4856. But that's not the case here, and we need those files to be processed. The only way to do that is to use m4 to process these files into *.s files which can then be given to Bazel's cc-* rules (its placed in the same zip file above).

Another issue is how GMP compiles many of these generated *.c files. They require macros to be passed to the compiler, which is different for every single file. Else the file cannot be compiled. This may not have been much of an issue, but with Tree Artifacts, this seems a bit hard to solve. And I'm currently trying to solve this problem.

https://github.com/robin-thomas/bazelify-gmp

**Comment by prestonvanloon** [comment]
Thanks for the update @robin-thomas! Can [select](https://docs.bazel.build/versions/master/be/functions.html#select) help with choosing the correct files based on the architecture?

**Comment by robin-thomas** [comment]
@prestonvanloon No. That *select* is a bit different. It's like a dictionary lookup based on a key, where the key can be architecture or some other parameters, but the value stored in the dictionary should be pre-defined. That's not the case here since we don't know the contents of the dictionary beforehand.

Anyway, I think I might have found another way to tackle it. So trying that.

**Comment by robin-thomas** [comment]
Seems like I found a small bug in Bazel's code: https://github.com/bazelbuild/bazel/issues/6963

Maybe I'll wait for someone from their dev team to reply there.

**Comment by prestonvanloon** [comment]
Sounds good, that certainly looks like a bug to me.

**Comment by robin-thomas** [comment]
@prestonvanloon Yeah. It's a bit of a blocker for me. But I'll try another way to continue working.

Congrats on getting funds from Vitalik! Seems like it's making the news 👍

**Comment by robin-thomas** [comment]
@prestonvanloon Have hit the issue of https://github.com/bazelbuild/bazel/issues/1920

Basically, the issue is that Bazel does not do linking for *cc_library*. So if I create a *cc_library* for gmp, it would not include any of its transitive dependencies. But I can create a shared library using *cc_binary* (Bazel does linking here), but for some reasons, it fails the *go_library* in *go-bls* with error `expected cc_library, objc_library, cc_proto_library or cc_import`.

I have created the issue here: https://github.com/bazelbuild/rules_go/issues/1871

**Comment by robin-thomas** [comment]
Fixed it all. Just waiting for this issue to be cleared: https://github.com/prysmaticlabs/go-bls/issues/7#issuecomment-449609573

**Comment by robin-thomas** [comment]
@rauljordan @prestonvanloon 

I have made 2 PRs:

https://github.com/prysmaticlabs/go-bls/pull/8
This one creates a new `cc_library` called `go-lib` inside the BUILD file. This is to help the issue that I had mentioned in https://github.com/prysmaticlabs/go-bls/issues/7. Which means, `go-bls` can be kept clean without any mcl. But need to let Bazel create all the libraries. If you create another bounty for that, I would be happy to move mcl to Bazel. But for now, I'm not building it, and had just added my gmp as a dependency to `go-bls`.

Its not easy to integrate OpenSSL with Bazel. So I had included `BoringSSL` into the `cc_library`, if you want to go with the above route.

This has a directory called `external` with a BUILD file. Bazel currently doesnt support transitive WORKSPACE. That means, everything in my WORKSPACE, should be written in `go-bls` WORKSPACE, and that goes for any projects that depends on `go-bls`.

https://github.com/prysmaticlabs/mcl/pull/1
This one exports a few header files to help move out mcl from `go-bls`.

My repo is: https://github.com/robin-thomas/bazelify-gmp

Here is the output after running `go-bls` unit tests (with bazel gmp):
```
$ bazel test ...
Starting local Bazel server and connecting to it...
INFO: Invocation ID: df65dd4b-99e0-4542-975c-8dd574e59b74
INFO: SHA256 (https://codeload.github.com/golang/tools/zip/3e7aa9e59977626dc60433e9aeadf1bb63d28295) = 2384fa91351a7414b643c5230422ce45f5aa2be8a82727609afd4e64e6973a30
INFO: Repository rule 'bazelify_gmp' returned: {"remote": "https://github.com/robin-thomas/bazelify-gmp", "commit": "3819642a7f67cb6cd001edb862079238eb6d0a18", "shallow_since": "2018-12-23", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "bazelify_gmp"}
INFO: Repository rule 'boringssl' returned: {"remote": "https://github.com/google/boringssl", "commit": "fafc4482e85c09e7af5f71b2eb287b73ccd1020a", "shallow_since": "2018-12-21", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "boringssl"}
INFO: Analysed 3 targets (34 packages loaded, 9252 targets configured).
INFO: Found 2 targets and 1 test target...
INFO: Elapsed time: 63.900s, Critical Path: 35.70s
INFO: 623 processes: 623 linux-sandbox.
INFO: Build completed successfully, 657 total actions
//:go_default_test                                                       PASSED in 0.2s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 657 total actions
```

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 1000.0 DAI (1000.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@robin-thomas](https://gitcoin.co/profile/robin-thomas)



@ceresstation please take a look at the submitted work:
* [PR](https://github.com/prysmaticlabs/go-bls/pull/10) by @robin-thomas

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $52,632.56 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by robin-thomas** [comment]
@prestonvanloon Fixed by PR: https://github.com/prysmaticlabs/go-bls/pull/10
If everything is okay, can you kindly release the bounty?

**Comment by prestonvanloon** [comment]
@robin-thomas Thanks for all of your effort. 
I'd like to confirm this builds on Mac OS X and Windows, then we can release the bounty!

**Comment by robin-thomas** [comment]
@prestonvanloon Definitely. Let me know if you face any issues with building on Mac OS X and Windows. I'll wait for confirmation. Thanks!

**Comment by terencechain** [comment]
`bazel test //...` on my mac failed on the following:
https://gist.github.com/terenc3t/bdbdadf215f7df4dc1b2067e0bd7120b

Lmk if I'm missing anything

**Comment by robin-thomas** [comment]
@terenc3t Let me try reproducing it on my Mac.

**Comment by robin-thomas** [comment]
I can reproduce the issue in Mac. Let me try fixing it.

**Comment by robin-thomas** [comment]
@terenc3t Can you try out https://github.com/prysmaticlabs/go-bls/pull/12 in your Mac and see? It's working on my Mac.

**Comment by terencechain** [comment]
@robin-thomas confirmed #12 works on my Mac

**Comment by rauljordan** [comment]
Hey @robin-thomas, tried out bazel test and received this: any ideas?
```
$ bazel test //...
INFO: Analysed 3 targets (0 packages loaded).
INFO: Found 2 targets and 1 test target...
ERROR: /home/zypherx/.cache/bazel/_bazel_zypherx/6dc90437243ef5aea30a94d0f60323cd/external/gmp_6_1_2/BUILD.bazel:310:1: Executing genrule @gmp_6_1_2//:gen_mpn_objs failed (Exit 1)
cp: cannot create regular file 'external/gmp_6_1_2/mpn/fac_table.h': Permission denied
INFO: Elapsed time: 0.355s, Critical Path: 0.04s
INFO: 0 processes.
FAILED: Build did NOT complete successfully

FAILED: Build did NOT complete successfully
```

**Comment by robin-thomas** [comment]
@rauljordan Whats the chmod permission of the directory `/home/zypherx/.cache/bazel/_bazel_zypherx/6dc90437243ef5aea30a94d0f60323cd/external/gmp_6_1_2/mpn`?

**Comment by rauljordan** [comment]
```bash
$ ls -ld /home/zypherx/.cache/bazel/_bazel_zypherx/6dc90437243ef5aea30a94d0f60323cd/external/gmp_6_1_2/mpn
drwxrwxr-x 28 zypherx zypherx 20480 Dec 28 09:07 /home/zypherx/.cache/bazel/_bazel_zypherx/6dc90437243ef5aea30a94d0f60323cd/external/gmp_6_1_2/mpn
```

**Comment by rauljordan** [comment]
Tried an expunge before building and testing - build succeeds, tests fail as above

**Comment by robin-thomas** [comment]
Build succeeds, but the test fails? That's weird. The `cp` command is in the build stage. So both should trigger it. Will it be possible for you to run the bazel test as root user?

Which machine are you on, by the way? I'm unable to reproduce this error in both Ubuntu and Mac.

**Comment by prestonvanloon** [comment]
Windows fails to build
```
$ bazel test //...
ERROR: Failed to query DisplayName of HKCU\Software\Microsoft\Windows\CurrentVersion\Uninstall\cef613b089ee1c28a22489bde1c8f7f5
INFO: Invocation ID: 2ecdbd45-1b0e-44f4-813f-0fee5da89e5d
Loading:
Loading: 0 packages loaded
ERROR: error loading package '': Encountered error while reading extension file 'm4.bzl': no such package '@io_bazel_rules_m4//': Traceback (most recent call last):
        File "C:/users/$USER/_bazel_$USER/wrsgbola/external/bazel_tools/tools/build_defs/repo/git.bzl", line 170
                _clone_or_update(ctx)
        File "C:/users/$USER/_bazel_$USER/wrsgbola/external/bazel_tools/tools/build_defs/repo/git.bzl", line 73, in _clone_or_update
                fail(("error cloning %s:\n%s" % (ctx....)))
error cloning io_bazel_rules_m4:
+ cd C:/users/$USER/_bazel_$USER/wrsgbola/external
+ rm -rf C:/users/$USER/_bazel_$USER/wrsgbola/external/io_bazel_rules_m4 C:/users/$USER/_bazel_$USER/wrsgbola/external/io_bazel_rules_m4
      1 [main] rm (5048) C:\Program Files\Git\usr\bin\rm.exe: *** fatal error - cygheap base mismatch detected - 0x1802FE408/0x180301408.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version should
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
```

Note: username replaced by `$USER` for copy/paste.

The steps followed were simply to install bazel, git clone the repo, and `bazel test //...`, but this looks like it might be an issue with their environment...

**Comment by robin-thomas** [comment]
@prestonvanloon Shouldn't it be `$ bazel test //...`?'

The output shows `bazel test`. That shouldn't build anything

**Comment by prestonvanloon** [comment]
@robin-thomas I am actually getting the same issue as @rauljordan.

I built and tested everything yesterday. Then when I tried to build the test again (I expected the build to be cached so nothing happen), it tries to build gmp again and fails.

```
INFO: Elapsed time: 125.146s, Critical Path: 41.33s
INFO: 691 processes: 689 linux-sandbox, 2 local.
INFO: Build completed successfully, 849 total actions
//:go_default_test                                                       PASSED in 0.2s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warninINFO: Build completed successfully, 849 total actions
➜  go-bls git:(master) bazel build :go_default_test
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 4276e43b-c7b5-4c39-bba4-cd4b74878cbd
INFO: Analysed target //:go_default_test (38 packages loaded, 9962 targets configured).
INFO: Found 1 target...
ERROR: /home/preston/.cache/bazel/_bazel_preston/0b01e4150831f4fda0c858be1b4e17f3/external/gmp_6_1_2/BUILD.bazel:310:1: Executing genrule @gmp_6_1_2//:gen_mpn_objs failed (Exit 1)
cp: cannot create regular file 'external/gmp_6_1_2/mpn/fac_table.h': Permission denied
Target //:go_default_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 19.724s, Critical Path: 14.56s
INFO: 1 process: 1 local.
FAILED: Build did NOT complete successfully
```

@robin-thomas `bazel test //...` is the canonical version of "Build and run all tests from the workspace root" so yes this is the correct command.

**Comment by robin-thomas** [comment]
@prestonvanloon No I mean, the windows output shows `$ bazel test`. It should be `$ bazel test //...` or `$ bazel test ...`

**Comment by prestonvanloon** [comment]
oh, copy and paste error. Let me edit the comment

**Comment by prestonvanloon** [comment]
How to reproduce the issue from @rauljordan 

```
bazel clean --expunge
bazel build :go_default_test # Builds fine
bazel build :go_default_test # Second build fails (it shouldn't be building again anyway)
```

**Comment by robin-thomas** [comment]
@prestonvanloon The Windows issue, I'm not sure. Maybe it's a Cygwin issue. Can you try using some other shell? (Shouldn't you be able to run bazel from `cmd` or maybe try `git bash`?)

For the other issue, It shouldn't be building the second time anyway. Unless the genrule isn't caching. In which case, the whole build needs to be done again.

Seems like I can reproduce the error when I build the second time. Let me check it out.

**Comment by prestonvanloon** [comment]
@robin-thomas I'll follow up a bit more on the windows issue. I don't think bazel was installed correctly.

**Comment by robin-thomas** [comment]
@prestonvanloon I had installed Bazel using the Windows instructions and it didn't require Cygwin. You can install the bash shell. Plus the error seems to suggest an issue with Cygwin.

**Comment by robin-thomas** [comment]
@prestonvanloon @rauljordan Regarding the issue of build failing the second time, I had made the fix here: https://github.com/prysmaticlabs/go-bls/pull/13. It's working on my system. Can you try it out?

**Comment by robin-thomas** [comment]
@prestonvanloon I assume everything is working as expected on Mac and Linux?

Any updates regarding Windows?

**Comment by rauljordan** [comment]
Hey @robin-thomas tried it out on a fresh windows configuration and received this:
```
ERROR: C:/users/maiji/downloads/go-bls/BUILD.bazel:3:1: in _cgo_collect_info rule //:go_default_library%linux_386%cgo_info:
Traceback (most recent call last):
        File "C:/users/maiji/downloads/go-bls/BUILD.bazel", line 3
                _cgo_collect_info(name = 'go_default_library%linux_386%cgo_info')
        File "C:/users/maiji/_bazel_maiji/tufpbscd/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 414, in _cgo_collect_info_impl
                _CgoInfo(orig_srcs = ctx.files.srcs, transf..., <6 more arguments>)
        File "C:/users/maiji/_bazel_maiji/tufpbscd/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 421, in _CgoInfo
                _select_archives(ctx.attr.libs)
        File "C:/users/maiji/_bazel_maiji/tufpbscd/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 134, in _select_archives
                fail("cc_library(s) did not produce a...")
cc_library(s) did not produce any files
ERROR: Analysis of target '//:go_default_test' failed; build aborted: Analysis of target '//:go_default_library%linux_386%cgo_info' failed; build aborted
INFO: Elapsed time: 436.371s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (38 packages loaded, 7434 targets configured)
FAILED: Build did NOT complete successfully (38 packages loaded, 7434 targets configured)
```
However I'm not convinced this is an issue with the project but perhaps with how I set everything up - perhaps there was some other system dep I was missing

**Comment by robin-thomas** [comment]
@rauljordan Were you able to build `go-bls` in Windows before I made my first PR?

Maybe checkout to commit `ed485f9efa247225c65b29c1ee00b5b4590fdf38` and run a clean Bazel build (but modify it to use `@herumi_mcl` to be built in a `cc_library` rather than from the mcl directory inside go-bls)?

**Comment by robin-thomas** [comment]
@prestonvanloon Can I assume its working on Linux and Mac?

**Comment by rauljordan** [comment]
Hi @robin-thomas, confirmed it works on Linux and Mac. Waiting on a windows confirmation as I no longer have access to a windows machine to attempt. Perhaps @IvanTheGreatDev can help with this?

**Comment by robin-thomas** [comment]
@rauljordan That's great!
@IvanTheGreatDev Waiting to hear from you.

**Comment by rauljordan** [comment]
Hey @robin-thomas tried on another windows machine and received the same error. Checked out to that commit and then received this Java error

```
INFO: Invocation ID: e825c8f2-4267-4166-b7c9-1e5db3fae0f5
Loading:
Loading: 0 packages loaded
Analyzing: 3 targets (1 packages loaded, 0 targets configured)
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node '//:go_default_library%linux_386%_cgo_.o BuildConfigurationValue.Key[1ad0148498d0d593acb92abfdb3367c7] false' (requested by nodes '//:go_default_library%linux_386%cgo_import BuildConfigurationValue.Key[1ad0148498d0d593acb92abfdb3367c7] false')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:499)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:368)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source
```

Sorry this is taking so long to debug - I know also @JerryFireman has windows and can try this out.

**Comment by robin-thomas** [comment]
@rauljordan That's weird. That commit was before I even started working on gmp! So you had never built `go-bls` on Windows before?

**Comment by robin-thomas** [comment]
@JerryFireman @IvanTheGreatDev Have either of you built `go-bls` on Windows before?

**Comment by robin-thomas** [comment]
@rauljordan Regarding the first error, I had a look into cgo code.
https://github.com/bazelbuild/rules_go/blob/master/go/private/rules/cgo.bzl#L134

```def _select_archive(files):
    """Selects a single archive from a list of files produced by a
    static cc_library.
    In some configurations, cc_library can produce multiple files, and the
    order isn't guaranteed, so we can't simply pick the first one.
    """

    # list of file extensions in descending order or preference.
    exts = [".pic.lo", ".lo", ".a"]     <--- maybe need a windows specific extension here
    for ext in exts:
        for f in as_iterable(files):
            if f.basename.endswith(ext):
                return f
```
Windows static libraries have an extension *.lib (maybe Bazel produces *.a libraries on Windows, but I don't know. Can someone check this out?). If that's the case, then `cc_library` won't be able to find the library. Hence the error.

**Comment by robin-thomas** [comment]
@rauljordan Also, can you share the details about the Windows setup? What's the compiler used? If its MSVC, then `cgo` doesn't have support for it.

**Comment by rauljordan** [comment]
Hey Robin, it is indeed MSVC. That must be the issue in that case. @houjieth agreed to try it oit on windows this evening as well.

**Comment by robin-thomas** [comment]
@rauljordan Yup. That should be the issue then.
`cgo` does support MSYS, but I don't know to what extent. `cgo` support for Windows seems a bit poor.

So you never tested on Windows before?

@houjieth Try MSYS + gcc. Or maybe Git Bash (not sure whether a compiler is built into it)

**Comment by robin-thomas** [comment]
@houjieth Any updates?

**Comment by houjieth** [comment]
I'm using native bazel on windows (no cygwin, no mingw, no msys, no other simulated unix environment such as WSL).

Here's my output

```
C:\Users\houji\go\src\github.com\prysmaticlabs\go-bls>bazel build //...
Starting local Bazel server and connecting to it...
INFO: Repository rule 'io_bazel_rules_m4' returned: {"remote": "https://github.com/jmillikin/rules_m4", "commit": "2bf69df77dfb6b3ba6b7fc95c304b0dc279375bc", "shallow_since": "2018-12-22", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "io_bazel_rules_m4"}DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: 'BAZEL_VC' is not set, start looking for the latest Visual C++ installed.
DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Looking for VS%VERSION%COMNTOOLS environment variables, eg. VS140COMNTOOLS
DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Looking for Visual C++ through registry
DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Looking for default Visual C++ installation directory
DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: 'PROGRAMFILES(X86)' environment variable is not set, using 'C:\Program Files (x86)' as default
DEBUG: C:/users/houji/_bazel_houji/6dp3edvb/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Visual C++ build tools found at C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC
INFO: SHA256 (https://codeload.github.com/golang/tools/zip/3e7aa9e59977626dc60433e9aeadf1bb63d28295) = 2384fa91351a7414b643c5230422ce45f5aa2be8a82727609afd4e64e6973a30
INFO: Repository rule 'bazelify_gmp' returned: {"remote": "https://github.com/robin-thomas/bazelify-gmp", "commit": "98ccf38c714ccd5527abdb46c16095fa1f622070", "shallow_since": "2018-12-24", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "bazelify_gmp"}
INFO: Repository rule 'herumi_mcl' returned: {"remote": "https://github.com/prysmaticlabs/mcl", "commit": "b90f1fbfea78c47012fcbc9243a616e654794421", "shallow_since": "2018-12-27", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "herumi_mcl"}
INFO: Repository rule 'boringssl' returned: {"remote": "https://github.com/google/boringssl", "commit": "fafc4482e85c09e7af5f71b2eb287b73ccd1020a", "shallow_since": "2018-12-21", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [], "patch_tool": "patch", "patch_args": ["-p0"], "patch_cmds": [], "name": "boringssl"}
ERROR: C:/users/houji/go/src/github.com/prysmaticlabs/go-bls/BUILD.bazel:3:1: in _cgo_collect_info rule //:go_default_library%android_amd64%cgo_info:
Traceback (most recent call last):
        File "C:/users/houji/go/src/github.com/prysmaticlabs/go-bls/BUILD.bazel", line 3
                _cgo_collect_info(name = 'go_default_library%android_amd64%cgo_info')
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 414, in _cgo_collect_info_impl
                _CgoInfo(orig_srcs = ctx.files.srcs, transf..., <6 more arguments>)
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 421, in _CgoInfo
                _select_archives(ctx.attr.libs)
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 134, in _select_archives
                fail("cc_library(s) did not produce a...")
cc_library(s) did not produce any files
ERROR: Analysis of target '//:go_default_test' failed; build aborted: Analysis of target '//:go_default_library%android_amd64%cgo_info' failed; build aborted
INFO: Elapsed time: 433.229s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (35 packages loaded, 7453 targets configured)
```

I'm going to try MinGW + MSYS next.

**Comment by robin-thomas** [comment]
@houjieth From the log, it seems to be using MSVC. If that's the case, then the error is expected.

**Comment by houjieth** [comment]
Tried on MSYS2, got the same error. How can I enforce it to use gcc instead of MSVC? I'm really not familiar with MinGW and MSYS2...

**Comment by robin-thomas** [comment]
Are you able to run the below command?

```
$ bazel build //... --compiler=mingw-gcc
```

@houjieth Is there somewhere we can chat?

Also run the below command, and say whats the output:

```
$ bazel build @bazelify_gmp//:gmp
```

**Comment by houjieth** [comment]
Now I'm using MSYS2 now without MSVC.

```
houji@DESKTOP-O5H7M9A MSYS /c/Users/houji/go/src/github.com/prysmaticlabs/go-bls
$ bazel build //...
INFO: Invocation ID: 4ce4bc40-5265-454a-b469-00a9509a87e9
Loading:
Loading: 0 packages loaded
Analyzing: 3 targets (0 packages loaded, 0 targets configured)
ERROR: C:/users/houji/go/src/github.com/prysmaticlabs/go-bls/BUILD.bazel:3:1: in _cgo_collect_info rule //:go_default_library%linux_mipsle%cgo_info:
Traceback (most recent call last):
        File "C:/users/houji/go/src/github.com/prysmaticlabs/go-bls/BUILD.bazel", line 3
                _cgo_collect_info(name = 'go_default_library%linux_mipsle%cgo_info')
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 414, in _cgo_collect_info_impl
                _CgoInfo(orig_srcs = ctx.files.srcs, transf..., <6 more arguments>)
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 421, in _CgoInfo
                _select_archives(ctx.attr.libs)
        File "C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_go/go/private/rules/cgo.bzl", line 134, in _select_archives
                fail("cc_library(s) did not produce a...")
cc_library(s) did not produce any files
ERROR: Analysis of target '//:go_default_test' failed; build aborted: Analysis of target '//:go_default_library%linux_mipsle%cgo_info' failed; build aborted
INFO: Elapsed time: 0.592s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
```

```
houji@DESKTOP-O5H7M9A MSYS /c/Users/houji/go/src/github.com/prysmaticlabs/go-bls
$ bazel build @bazelify_gmp//:gmp
INFO: Invocation ID: 18ac6ddd-788d-4f38-b1f5-edd31178d40d
Loading:
Loading: 0 packages loaded
Analyzing: target @bazelify_gmp//:gmp (0 packages loaded, 0 targets configured)
INFO: Analysed target @bazelify_gmp//:gmp (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
[0 / 21] no action
ERROR: C:/users/houji/_bazel_houji/6dp3edvb/external/m4_v1.4.18/BUILD.bazel:15:1: C++ compilation of rule '@m4_v1.4.18//:m4_lib' failed (Exit 2)
c:\users\houji\_bazel_houji\6dp3edvb\execroot\__main__\external\m4_v1.4.18\lib\freadahead.h(41): error C2061: syntax error: identifier '__attribute__'
c:\users\houji\_bazel_houji\6dp3edvb\execroot\__main__\external\m4_v1.4.18\lib\freadahead.h(41): error C2059: syntax error: ';'
Target @bazelify_gmp//:gmp failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.788s, Critical Path: 0.70s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

If add `--compiler=mingw-gcc`

```
houji@DESKTOP-O5H7M9A MSYS /c/Users/houji/go/src/github.com/prysmaticlabs/go-bls
$ bazel build //... --compiler=mingw-gcc
INFO: Invocation ID: dc0ee7fe-a156-442e-b322-d57dca3822ed
Loading:
Loading: 0 packages loaded
INFO: Build option --compiler has changed, discarding analysis cache.
Analyzing: 3 targets (0 packages loaded, 0 targets configured)
INFO: Analysed 3 targets (0 packages loaded, 7846 targets configured).
INFO: Found 3 targets...
[4 / 16] [-----] BazelWorkspaceStatusAction stable-status.txt
ERROR: C:/users/houji/_bazel_houji/6dp3edvb/external/io_bazel_rules_m4/m4/internal/BUILD:1:1: C++ compilation of rule '@io_bazel_rules_m4//m4/internal:capture_stdout' failed (Exit 1)
external/io_bazel_rules_m4/m4/internal/capture_stdout.c: In function 'main':
external/io_bazel_rules_m4/m4/internal/capture_stdout.c:31:48: error: 'S_IRUSR' undeclared (first use in this function)
     out_fd = open(argv[1], O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
                                                ^~~~~~~
external/io_bazel_rules_m4/m4/internal/capture_stdout.c:31:48: note: each undeclared identifier is reported only once for each function it appears in
external/io_bazel_rules_m4/m4/internal/capture_stdout.c:31:58: error: 'S_IWUSR' undeclared (first use in this function)
     out_fd = open(argv[1], O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
                                                          ^~~~~~~
INFO: Elapsed time: 2.543s, Critical Path: 0.37s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

```
houji@DESKTOP-O5H7M9A MSYS /c/Users/houji/go/src/github.com/prysmaticlabs/go-bls
$ bazel build @bazelify_gmp//:gmp --compiler=mingw-gcc
INFO: Invocation ID: 768ff9c7-3663-4166-a1a0-24d2444af630
Loading:
Loading: 0 packages loaded
Analyzing: target @bazelify_gmp//:gmp (0 packages loaded, 0 targets configured)
INFO: Analysed target @bazelify_gmp//:gmp (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
[0 / 17] no action
ERROR: C:/users/houji/_bazel_houji/6dp3edvb/external/m4_v1.4.18/BUILD.bazel:15:1: C++ compilation of rule '@m4_v1.4.18//:m4_lib' failed (Exit 1)
In file included from external/m4_v1.4.18/lib/regex.c:68:
external/m4_v1.4.18/lib/regex_internal.h:29:10: fatal error: langinfo.h: No such file or directory
 #include <langinfo.h>
          ^~~~~~~~~~~~
compilation terminated.
Target @bazelify_gmp//:gmp failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.267s, Critical Path: 0.21s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Comment by robin-thomas** [comment]
Thanks for the logs @houjieth!
Can you check whether the Windows platform has the `m4` binary in `PATH`?

@prestonvanloon It seems like `rules_m4` doesn't work on Windows. So what to do now?

**Comment by rauljordan** [comment]
Hey @robin-thomas, m4 is here on sourceforge http://gnuwin32.sourceforge.net/packages/m4.htm, and perhaps we can use bazel platform selection to download the right distribution? It seems to be downloading the linux version with the current configuation. @prestonvanloon suggested this.

**Comment by robin-thomas** [comment]
@rauljordan @prestonvanloon I had fixed all the m4 issues, and other issues that came later. It now builds GMP library fine. But `herumi_mcl` build fails. Have you tried building `go_bls` or `herumi_mcl` on windows before?

**Comment by rauljordan** [comment]
@houjieth can you give this another try?

**Comment by 0xKiwi** [comment]
```
ivanthegreat@DESKTOP-T8PCO20:/mnt/n/go-bls$ bazel build ...
Starting local Bazel server and connecting to it...
INFO: Analysed 3 targets (37 packages loaded).
INFO: Found 3 targets...
INFO: From Compiling external/m4_v1.4.18/lib/c-stack.c [for host]:
external/m4_v1.4.18/lib/c-stack.c:107:1: warning: 'die' defined but not used [-Wunused-function]
 die (int signo)
 ^~~
INFO: Elapsed time: 161.439s, Critical Path: 138.57s
INFO: 669 processes: 2 local, 667 processwrapper-sandbox.
INFO: Build completed successfully, 678 total actions
```

Builds on windows!

**Comment by rauljordan** [comment]
The above works on Hyper, but not on MSYS. However, this is good enough for us as a lot of work has already been put in. Allowing for support all across the board including MSYS can be a separate PR or bounty.

**Comment by rauljordan** [comment]
@ceresstation we are happy with the bounty - I can edit the README post merge to include details about building on Windows.

**Comment by spm32** [comment]
Fantastic to hear @rauljordan, thanks for all of your help @robin-thomas!

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 1000.0 DAI (1000.0 USD @ $1.0/DAI)  attached to this issue has been approved & issued to @robin-thomas.__  

 
* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/prysmaticlabs/go-bls/2/1782)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $42,781.51 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by luxe** [comment]
Just to clarify, is the single `bazelify_gmp` all that is necessary to start using gmp in bazel projects?
```
git_repository(
    name = "bazelify_gmp",
    remote = "https://github.com/robin-thomas/bazelify-gmp",
    commit = "bb4881b35e6864c90493980d035e1d984cafd093",
)
```

Or is it also required to bring in the m4 rules?

```
git_repository(
    name = "io_bazel_rules_m4",
    remote = "https://github.com/jmillikin/rules_m4",
    commit = "2bf69df77dfb6b3ba6b7fc95c304b0dc279375bc",
)
load("@io_bazel_rules_m4//:m4.bzl", "m4_register_toolchains")
```

m4 is a dependency of gmp, but I'm not sure if the bazelify-gmp is self-sufficient to use alone.  Thanks.

**Comment by prestonvanloon** [comment]
@luxe Yes, I think you would need to install the transitive dependencies.
